### PR TITLE
 feat(email): support e-mail attachments 

### PR DESCRIPTION
**Pull request by jeabakker**
The email service now supports attachments

- [x] docs
- [x] unit test

### CODE REVIEW

**Comment by jeabakker** [comment]
need a suggestion for the documentation location

File: engine/classes/Elgg/EmailService.php [hunk 1]
```diff
@@ -127,6 +122,19 @@ public function transport(Email $email) {
 			// which influences the output of the header in the message
 			$message->getHeaders()->addHeaderLine("{$name}: {$value}");
 		}
+		
+		// set Subject
+		$subject = elgg_strip_tags($email->getSubject());
+		$subject = html_entity_decode($subject, ENT_QUOTES, 'UTF-8');
```
Comment (hypeJunction) [inline engine/classes/Elgg/EmailService.php]: Please add tests for these changes

File: engine/classes/Elgg/EmailService.php [hunk 2]
```diff
@@ -142,5 +150,185 @@ public function transport(Email $email) {
...
+			if ($attachment instanceof Part) {
+				$part = $attachment;
+			}
+			
+			if ($attachment instanceof \ElggFile) {
+				$part = $this->elggFileToAttachement($attachment);
+			}
+			
+			if (is_array($attachment)) {
+				$part = $this->arrayToAttachement($attachment);
```
Comment (hypeJunction) [inline engine/classes/Elgg/EmailService.php]: Attachment not Attachement

File: engine/classes/Elgg/EmailService.php [hunk 3]
```diff
@@ -142,5 +150,185 @@ public function transport(Email $email) {
...
+	}
+	
+	/**
+	 * Make an ElggFile into a correct formatted attachment
+	 *
+	 * @param \ElggFile $file the ElggFile to process
+	 *
+	 * @return false|\Zend\Mime\Part
+	 */
+	protected function elggFileToAttachement(\ElggFile $file) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/EmailService.php]: I don't think this belongs here, make some utility/adapter class. Better yet add ElggEmailAttachment interface that extends Part and require that attachments comply with the instance. I don't like all this boilerplate here

**Review by hypeJunction — COMMENTED** [review]
Please finish the tests before merging. There is no rush to get these in and it's impossible to make further refactor when tests are incomplete.

File: engine/classes/Elgg/EmailService.php [hunk 4]
```diff
@@ -142,5 +150,185 @@ public function transport(Email $email) {
...
 	}
+	
+	/**
+	 * Build the body part of the e-mail message
+	 *
+	 * @param Email $email Email
+	 *
+	 * @return \Zend\Mime\Message
+	 */
+	protected function buildMessageBody(Email $email) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/EmailService.php]: Why not just attach files directly to Email instance?

File: engine/classes/Elgg/EmailService.php (hunk 1 above)
Comment (jeabakker) [reply engine/classes/Elgg/EmailService.php]: this is not a change, this has always been the case

File: engine/classes/Elgg/EmailService.php (hunk 4 above)
Comment (jeabakker) [reply engine/classes/Elgg/EmailService.php]: could be, but than we'd have to maybe rewrite a lot of our current code to get it into the right place.

Now it works with all the current code in place, just add `attachments` to the `$params` array and it works

File: engine/classes/Elgg/EmailService.php (hunk 3 above)
Comment (jeabakker) [reply engine/classes/Elgg/EmailService.php]: Maybe in the Email class?

File: engine/classes/Elgg/EmailService.php (hunk 4 above)
Comment (hypeJunction) [reply engine/classes/Elgg/EmailService.php]: I just untangled 10 years of email patching. Save your future self some time and do it right from the start with all the tests in place.

File: engine/classes/Elgg/EmailService.php (hunk 3 above)
Comment (hypeJunction) [reply engine/classes/Elgg/EmailService.php]: I'd say add `EmailAttachment` class with static `factory` methods. Then you could just use `Email::attach(EmailAttachment)`. Somewhere along the lines, add a hook to filter attachments.

File: engine/classes/Elgg/EmailService.php (hunk 1 above)
Comment (jeabakker) [reply engine/classes/Elgg/EmailService.php]: ps: this is tested https://github.com/Elgg/Elgg/blob/master/engine/tests/phpunit/unit/Elgg/EmailServiceUnitTest.php#L28

**Comment by jeabakker** [comment]
made some changed, added docs and tests. ready for re-review

**Comment by hypeJunction** [comment]
Looks much better. Thanks

**Pull request merged**